### PR TITLE
fix: prevent zero-probability events from being sampled in multinomial function in Jax-backend

### PR DESCRIPTION
**Pull request by SinanGncgl**
## Description

This PR fixes the issue #7 where zero-probability events could still be sampled due to the use of a small constant (1e-37) in the logits calculation. The fix replaces this constant with -inf to ensure such events are never selected and eliminate zero prob event.

## What is Changed

Include the changes introduced in this PR.

- Updated the multinomial implementation to handle zero-probability outcomes correctly.
- Replaced 1e-37 with -inf using jnp.where for stable log calculations.

## Checklist:

- [x] Tests that cover the code added.
- [x] Corresponding changes documented.
- [x] All tests passed.
- [x] The code linted and styled (pre-commit run --all-files has passed).

### CODE REVIEW

File: mithril/backends/with_autograd/jax_backend/backend.py [hunk 1]
```diff
@@ -513,29 +513,30 @@ def multinomial(
...
         prng_key = self._get_prng_key(key)
-        input = jax.numpy.asarray(probs)
-        input = input / jax.numpy.sum(input, axis=-1, keepdims=True)
-        batch_size = input.shape[:-1]
-        logits = jax.numpy.log(jax.numpy.maximum(input, 1e-37))
+        probs = jnp.asarray(probs)
+        probs = probs / jnp.sum(probs, axis=-1, keepdims=True)
+
+        # Mask zero probabilities to avoid log(0) without adding small constants
+        logits = jnp.where(probs > 0, jnp.log(probs), -jnp.inf)
```
Comment (mehmetozsoy-synnada) [inline mithril/backends/with_autograd/jax_backend/backend.py:520]: Instead of using `jnp.where` in order to avoid numerical instabilities, I tried this 

```python3
logits = jax.numpy.log(probs)
```

I tested the edge cases in simple ways and did not encounter any problems as far as I observed. I think we could also write it this way, as it is simpler.

File: mithril/backends/with_autograd/jax_backend/backend.py [hunk 2]
```diff
@@ -513,29 +513,30 @@ def multinomial(
...
 
             # For without replacement, use Gumbel-max trick
             # This is much faster than using choice
-            z = jax.random.gumbel(prng_key, shape=input.shape + (num_samples,))
+            z = jax.random.gumbel(prng_key, shape=probs.shape + (num_samples,))
             # Add log probabilities for Gumbel-max trick,
             z = z + logits[..., None]
             # Get top k indices
-            samples = jax.numpy.argsort(-z, axis=input.ndim - 1)[..., :num_samples, 0]
+            samples = jax.numpy.argsort(-z, axis=probs.ndim - 1)[..., :num_samples, 0]
```
Comment (mehmetozsoy-synnada) [inline mithril/backends/with_autograd/jax_backend/backend.py:539]: `Codecov` checks in CI are failing because this function is not currently tested. Relevant tests can be added. Backend functions are often tested in the `test_backend_fns.py` file, so multinomial tests can be added there.

File: mithril/backends/with_autograd/jax_backend/backend.py (hunk 1 above)
Comment (ozankabak) [reply mithril/backends/with_autograd/jax_backend/backend.py:520]: Seems like the `log` function automatically returns `-inf` when the input is zero.

File: mithril/backends/with_autograd/jax_backend/backend.py (hunk 2 above)
Comment (ozankabak) [reply mithril/backends/with_autograd/jax_backend/backend.py:539]: @SinanGncgl, let's add unit test(s) for this in `test_backend_fns.py` file and try the simplified expression above (without the `where`), then we can merge your PR! Thank you for your contribution 🚀

Comment (SinanGncgl) [reply mithril/backends/with_autograd/jax_backend/backend.py:539]: Thanks for the suggestions, I will make the necessary changes and commit again.

**Review by mehmetozsoy-synnada — APPROVED** [review]

**Comment by ozankabak** [comment]
Seems like we have a formatting problem, but it otherwise looks great. If you make sure the pre-commit procedure passes, we can proceed with the merge.

File: tests/scripts/test_backend_fns.py [hunk 3]
```diff
@@ -2060,3 +2060,42 @@ def test_rand_uniform(self, backendcls, device, dtype):
...
+        assert not backend.any(output < 0)
+        assert not backend.any(output >= len(fn_args[0]))
+        assert list(output.shape) == [fn_args[1]]
+
+    def test_multinomial_invalid_probabilities(self, backendcls, device, dtype):
+        backend = backendcls(device=device, dtype=dtype)
+        fn = backend.multinomial
+
+        array_size = 1000
+        fn_args = [backend.array([0.0] * 500 + [1.0] + [0.0] * (array_size - 501)), 100]  # Sample 100 times
```
Comment (mehmetozsoy-synnada) [inline tests/scripts/test_backend_fns.py:2096]: ruff checks in CI pre-commit are failing because this line is too long, but other than that, LGTM.

File: mithril/backends/with_autograd/jax_backend/backend.py [hunk 4]
```diff
@@ -513,29 +513,28 @@ def multinomial(
             replacement: whether to sample with replacement
         """
         prng_key = self._get_prng_key(key)
-        input = jax.numpy.asarray(probs)
-        input = input / jax.numpy.sum(input, axis=-1, keepdims=True)
-        batch_size = input.shape[:-1]
-        logits = jax.numpy.log(jax.numpy.maximum(input, 1e-37))
+        probs = jnp.asarray(probs)
+        probs = probs / jnp.sum(probs, axis=-1, keepdims=True)
+        logits = jnp.log(probs)
```
Comment (ozankabak) [inline mithril/backends/with_autograd/jax_backend/backend.py:518]:
```suggestion
        probs = probs / jnp.sum(probs, axis=-1, keepdims=True)
        logits = jnp.log(probs)
```

File: mithril/backends/with_autograd/jax_backend/backend.py [hunk 5]
```diff
@@ -513,29 +513,28 @@ def multinomial(
...
 
         if replacement:
             # Use categorical directly - much faster than choice
             samples = jax.random.categorical(
                 prng_key,
                 logits,  # avoid log(0)
```
Comment (ozankabak) [inline mithril/backends/with_autograd/jax_backend/backend.py:524]:
```suggestion
                logits,
```

**Review by ozankabak — APPROVED** [review]

**Comment by ozankabak** [comment]
Thank you @SinanGncgl 🚀
